### PR TITLE
Addition of ‘--check’ option for ‘upload’ and ‘sdist’

### PR DESCRIPTION
**Pull request by mrkkrp**
See #1183. Here I've added `--check` switch. Previously simple tuple was used to hold options for `upload` and `sdist` commands. Now with the addition of `--check` option, I switched to `(,,)`. Is it OK? Should some data structure be defined instead of tuple? This is very beginning of my work, so I create this PR early to ask questions along the way.

### CODE REVIEW

**Comment by mrkkrp** [comment]
Is it better to throw exception to stop current action from checking function or from its caller (in `sdistCmd` or `uploadCmd` i.e. depending on results of check)? Checking function can be coded to return collection of `PackageCheck` things, which then can be analyzed. Another option is to print warnings inside of this function and throw exception if necessary. What is preferable?

**Comment by mrkkrp** [comment]
I've done some progress, I guess. I do have more questions, without answers I won't be able to proceed (or it will take more time):

* how to get `PackageConfig`;
* how to get `PackageDescription`.

see `Stack.Check`.

Also please review what I already have.

**Comment by mgsloan** [comment]
Thanks for working on this!  Sorry for the delayed feedback.

RE the first question, I think the check should just throw exceptions.  I figure that it may as well throw, since at that point, execution isn't going to continue beyond letting the user know how it failed.

What you've got looks pretty good so far.  It might be tricky to get "checkPackageContent" to work properly, though.  Instead of doing existence checks based on the working tree, they should be based on the tarball contents.  This should be possible with the in-memory representation of the sdist, but maybe it's easier to just unpack it.

Here's how to get `PackageConfig`: https://github.com/commercialhaskell/stack/blob/15dda17fb1b42d6961caea95a0b24c493758c04c/src/Stack/SDist.hs#L156

Note the TODO - ideally instead of copying this code yet a third time it would be put in a utility somewhere.

`PackageDescription` isn't as straightforward, since our datatypes extract all the data we need from the cabal datatypes, and don't store the cabal datatypes.  It's already getting computed when doing `readPackageDir`, which ends up invoking [resolvePackage](https://github.com/commercialhaskell/stack/blob/09991d98ed034e036668721de966c199e2e40642/src/Stack/Package.hs#L180) on a `gpkg` value.  I think `resolvePackageDescription packageConfig gpkg` is what you want.

Considering that `readPackageDir` isn't actually used anywhere else, I'd just rename it to `readPackageDescriptionDir`, and have it return a `PackageDescription`.

**Comment by mrkkrp** [comment]
Thanks for the explanation, I think I will finish it in a couple of days now!

**Comment by mrkkrp** [comment]
*(Rebased on current master.)* Please check this out. It works, but it may be kinda too sensitive (here I added `-O2` flag for testing):

```
~/projects/programs/haskell/stack $ stack sdist --check
Checking package 'stack' for common mistakes
Package check reported the following errors:
'ghc-options: -O2' is rarely needed. Check that it is giving a real benefit and not just imposing longer compile times on your users.
```

You see, this aborts the whole thing because of this. I don't know where should be the “threshold” where failed checks start to terminate current operation. `upload` is not currently affected. In `checkPackage` I used existing directory tree (uncompressed) because it's logical to check files before the whole thing goes into archive (also easier). Should be possible to do the same for `upload` (?)

I think I made the changes and refactoring you mentioned, see for yourself.

**Comment by mgsloan** [comment]
Looking pretty good so far!  It looks like travis is failing build it due to a scope error, though.

Also, when reviewing the issue / PR initially, I misunderstood what the checks entail (I thought it was like #717) .  I think these checks would be a good thing to do by default.  How about instead adding a `--ignore-check` flag?  This would still do the check, but ignore if it fails.

Feel free to continue using exceptions for this (and catch them), but this would probably be a little cleaner if they're just returned.

Also, maybe this is getting a little nitpicky, I'll merge either way, but: for the most part we've tended towards rather large modules in stack.  How about just putting it in the SDist module?  It can be moved out if commands other than sdist / upload end up using it.

>  I don't know where should be the “threshold” where failed checks start to terminate current operation.

I'd say it should be the same threshold as hackage (I'm not sure what this is), and otherwise emit warnings.

> In checkPackage I used existing directory tree (uncompressed) because it's logical to check files before the whole thing goes into archive (also easier)

The issue with this is that the files in the existing directory tree aren't necessarily included in the archive.

> Should be possible to do the same for upload (?)

I'm definitely in favor of this!

So, in summary:

- [ ] Default to checking, have `--ignore-check` flag
- [ ] Return results instead of throwing?
- [ ] Move checkPackage into SDist?
- [ ] Same error threshold as hackage
- [ ] Check file existence based on existence within the tar
- [ ] Use it for upload

**Comment by mrkkrp** [comment]
Since other functions work with paths outside of tar, I think I will temporarily unpack archive, check it, and delete temporary directory. Is this acceptable?

**Comment by mgsloan** [comment]
Yup, that's acceptable!

Based on what I saw of cabal's API last time I looked, I think you might be able supply it with the "does this file exist" / "does this dir exist" predicates.

**Comment by mrkkrp** [comment]
OK, here it is. Currently we decompress archives into temporary directories that are deleted after checking. What's worse, in `upload` command, if we are given directories, we get archives as lazy byte strings, which makes me temporarily save them *and only then unpack*. I don't like this at all.

I thought, maybe I could do everything in memory. We could use `gzip` library to get in-memory tar archive and then there is [interface](http://haddock.stackage.org/nightly-2015-11-17/tar-0.4.2.2/Codec-Archive-Tar-Entry.html) (although I've never worked with it) to get files from it (it's a bit hairy of course). This would mean rewriting of all helpers to work in this environment (because normally they work in file system), which I'm afraid I cannot do right now.

All in all, this should do the trick for now, tell me if there is something wrong with it.

File: src/Stack/Check.hs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+-- | Check package for common mistakes.
```
Comment (mgsloan) [inline src/Stack/Check.hs]: This file shouldn't get added

File: src/main/Main.hs [hunk 2]
```diff
@@ -817,20 +820,23 @@ uploadCmd (args, mpvpBounds) go = do
...
-            liftIO $ forM_ files (canonicalizePath >=> Upload.upload uploader)
-            forM_ dirs $ \dir -> do
-                pkgDir <- parseAbsDir =<< liftIO (canonicalizePath dir)
-                (tarName, tarBytes) <- getSDistTarball mpvpBounds pkgDir
-                liftIO $ Upload.uploadBytes uploader tarName tarBytes
+    withBuildConfigAndLock go $ \_ -> do
+      uploader <- getUploader
+      unless ignoreCheck $
+        mapM_ (liftIO . canonicalizePath >=>
+               parseAbsFile              >=>
```
Comment (mgsloan) [inline src/main/Main.hs]: `liftIO . canonicalizePath >=> parseAbsFile` is the same as `Path.IO.parseRelAsAbsDir`

File: src/main/Main.hs [hunk 3]
```diff
@@ -817,20 +820,23 @@ uploadCmd (args, mpvpBounds) go = do
...
+               checkSDistTarball)
+              files
+      liftIO $ forM_ files (canonicalizePath >=> Upload.upload uploader)
+      unless (null dirs) $
+        forM_ dirs $ \dir -> do
+          pkgDir <- parseAbsDir =<< liftIO (canonicalizePath dir)
```
Comment (mgsloan) [inline src/main/Main.hs]: Same here, this can also be `parseRelAsAbsDir`

**Comment by mgsloan** [comment]
Great stuff, thanks for addressing all my points.  Here's all that's left: 

- [ ] Use `parseRelAsAbsDir`
- [ ] Remove `Stack/Check.hs`
- [ ] Use 4-space indentation - I'd actually merge it anyway since it's not like indentation is consistent throughout the codebase.  However, most code in stack does use 4 spaces.

**Comment by mrkkrp** [comment]
@mgsloan, Done. I've come to using 2-space indentation almost everywhere (except for Python). Maybe it's because I've been writing Common and Emacs Lisps for a while in the past. Another consideration is when you have 2-space indentation you can put more code on a line (if you don't want to exceed 80 character width) and levels are still very visible, but that's a holy-war topic :-)

**Comment by mgsloan** [comment]
LGTM, thanks!

File: src/Stack/SDist.hs [hunk 4]
```diff
@@ -246,6 +260,61 @@ dirsFromFiles dirs = Set.toAscList (Set.delete "." results)
...
+checkSDistTarball tarball = withTempTarGzContents tarball $ \pkgDir' -> do
+    pkgDir  <- (pkgDir' </>) `liftM`
+        (parseRelDir . FP.takeBaseName . FP.takeBaseName . toFilePath $ tarball)
+    --               ^ drop ".tar"     ^ drop ".gz"
+    cabalfp <- getCabalFileName pkgDir
+    name    <- parsePackageNameFromFilePath cabalfp
+    config  <- getPackageConfig name
+    (gdesc, pkgDesc) <- readPackageDescriptionDir config pkgDir
+    $logInfo $ "Checking package '" <>
+               (T.pack . P.unPackageName . P.pkgName . PD.package $ pkgDesc) <>
```
Comment (sjakobi) [inline src/Stack/SDist.hs]: The `unPackageName` accessor has existed only since Cabal-1.22.

I ran into this when I tried to build stack using the stack-7.8.yaml.

Maybe there should be an additional Travis build using the stack-7.8.yaml?

Comment (mgsloan) [inline src/Stack/SDist.hs]: Good point!  I've resolved this with https://github.com/commercialhaskell/stack/commit/44da1c1bdfa3fb6d0f62a54f05859d80b0efd34b and opened https://github.com/commercialhaskell/stack/issues/1398
